### ISSUE TITLE
Cannot use custom form in contrib.modeladmin with some fields excluded

### ISSUE THREAD

**Issue by theju** [torchbox/wagtail#2717]
Supposing we have the following model

``` python
class Author(models.Model):
    first_name = models.CharField(max_length=50)
    last_name = models.CharField(max_length=50)
```

and customize the form to exclude a field

``` python
class AuthorAdminForm(forms.ModelForm):
    class Meta:
        model = Author
        fields = ("first_name", )
```

and add a Wagtail ModelAdmin

``` python
from wagtail.contrib.modeladmin.views import EditView
from wagtail.contrib.modeladmin.options import ModelAdmin, modeladmin_register

class AuthorEditView(EditView):
    def get_form_class(self):
        return AuthorAdminForm

class AuthorModelAdmin(ModelAdmin):
    model = Author
    edit_view_class = AuthorEditView
    list_display = ("first_name", "last_name")

modeladmin_register(AuthorModelAdmin)
```

When accessing `/admin/author/author/edit/<id>/`, we get an error that reads:

```
u"Key 'last_name' not found in 'AuthorAdminForm'"
```

The same is true when using the `create_view_class` attribute as well.

Maybe related to #2712

**Comment by jimr** [comment]
Yes, this is related to #2712 - see that issue for the suggested approach to achieving this (it doesn't involve custom forms). Whether or not forms should be override-able is something I've asked on that issue and I think we just need some more guidance in the docs about which things we're familiar with from "vanilla" Django we can do with `modeladmin` and which we can't.

**Comment by ababic** [comment]
Hi both.

I have addressed this to some degree on #2712, but It makes sense to repeat some of what I said here, as the issue name and thread are more relevant.

Using custom forms in modeladmin isn't something that is supported right now. In time, I'm hoping we might be able to find a way around this, but I'm not yet familiar enough with Wagtail's 'edit handler' layer to know what would be involved. I'm hoping @gasman might be able to provide some guidance there?

**Comment by gasman** [comment]
When edit handlers are in use (as they are in ModelAdmin), the edit handler mechanism is responsible for generating a ModelForm with the appropriate fields. In this case, I think setting a `panels` definition on the model should do everything you need:

    class Author(models.Model):
        first_name = models.CharField(max_length=50)
        last_name = models.CharField(max_length=50)

        panels = [
            FieldPanel('first_name')
        ]

More complex form customisations, such as form validation, can be done by defining a `base_form_class`, as per http://docs.wagtail.io/en/v1.5.2/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms.

(@theju, I guess this answers your question on wagtail-developers about why ModelAdmin doesn't follow the Django Admin API - it's because ModelAdmin adopts Wagtail's own 'edit handlers' mechanism for constructing modelforms, which is fundamentally different from django-admin's approach. This also means - I suspect - that it's not really compatible with supporting a `get_form_class` method.)

**Comment by theju** [comment]
With respect to `base_form_class`, it becomes tricky to have the form and the model in separate files because of the recursive import issue and, having both in the same file feels unnatural.

Another question is why doesn't the wagtail edit handler not take into account the form's fields/exclude attribute? Specifically, https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailadmin/edit_handlers.py#L252-L258

That way handling custom forms in modeladmin (and elsewhere) would also be possible. Reusing existing forms and validation would be another advantage of this fix.

Doing any customization on the edit_handlers generally replicates a lot of the form's (and sometimes template's) functionality and in my opinion, a violation of DRY.

Would love to know what the core wagtail devs think.

**Comment by theju** [comment]
Since I brought up the issue of honouring the form's fields/exclude attribute, here is a quick patch that I wrote (almost all tests pass).

```
diff --git a/wagtail/wagtailadmin/edit_handlers.py b/wagtail/wagtailadmin/edit_handlers.py
index 93ba79b..f961a10 100644
--- a/wagtail/wagtailadmin/edit_handlers.py
+++ b/wagtail/wagtailadmin/edit_handlers.py
@@ -252,10 +252,17 @@ class BaseCompositeEditHandler(EditHandler):
     def __init__(self, instance=None, form=None):
         super(BaseCompositeEditHandler, self).__init__(instance=instance, form=form)
 
-        self.children = [
-            handler_class(instance=self.instance, form=self.form)
-            for handler_class in self.__class__.children
-        ]
+        self.children = []
+        for child in self.__class__.children:
+            if not getattr(child, "children", None) and getattr(child, "field_name", None):
+                if child.field_name in (self.form._meta.exclude or []) or \
+                   child.field_name not in (self.form._meta.fields or []):
+                    continue
+            self.children.append(child(instance=self.instance, form=self.form))
 
     def render(self):
         return mark_safe(render_to_string(self.template, {
```

**Comment by theju** [comment]
An updated patch where all current tests pass and also have ability for custom forms in modeladmin

```
diff --git a/wagtail/wagtailadmin/edit_handlers.py b/wagtail/wagtailadmin/edit_handlers.py
index 93ba79b..31a2340 100644
--- a/wagtail/wagtailadmin/edit_handlers.py
+++ b/wagtail/wagtailadmin/edit_handlers.py
@@ -252,10 +252,20 @@ class BaseCompositeEditHandler(EditHandler):
     def __init__(self, instance=None, form=None):
         super(BaseCompositeEditHandler, self).__init__(instance=instance, form=form)
 
-        self.children = [
-            handler_class(instance=self.instance, form=self.form)
-            for handler_class in self.__class__.children
-        ]
+        self.children = []
+        for child in self.__class__.children:
+            if not getattr(child, "children", None) and getattr(child, "field_name", None):
+                if self.form._meta.exclude:
+                    if child.field_name in self.form._meta.exclude:
+                        continue
+                if self.form._meta.fields:
+                    if child.field_name not in self.form._meta.fields:
+                        continue
+            self.children.append(child(instance=self.instance, form=self.form))
 
     def render(self):
         return mark_safe(render_to_string(self.template, {
```

**Comment by ababic** [comment]
Hi @theju. Thanks for taking some time to investigate. Would you be able to add your changes as pull request and include your tests etc, so that it can be reviewed properly?

**Comment by theju** [comment]
PR #2752 has been created for this ticket.

**Comment by gasman** [comment]
Implemented in #2752.
